### PR TITLE
Wildcard support in launcher

### PR DESCRIPTION
**Pull request by nikunjy**
Supporting wildcard in kotest launcher based on the discussion here https://github.com/kotest/kotest/issues/2416#issuecomment-1233595843

I am extending the `testpath` to support wildcard. In the future we can  support regexes with very small changes. 

After this PR is merged you will be able to use the kotest-framework-engine-launcer with the following args
```
"--reporter", "teamcity", 
"--package", "com.kotlin.test",
"--spec", "*",
"--testpath", "db*",
```
The launcher already supports --testpath argument. And this PR is just extending that testpath option to be more flexible. It is written with the same reasoning in mind as the original --testpath arg.

Not only this but you would be able to use this to write a runner like this

```
 ExecuteKt.execute(
  reporter, 
  "<package name>",
  "<spec wildcard>"
  "<test wildcard>"
)
```

### CODE REVIEW

**Comment by nikunjy** [comment]
@sksamuel  what do you think. Wanna give me a code review ? :)

**Comment by LeoColman** [comment]
I think this PR is a duplicate of https://github.com/kotest/kotest/pull/3190. Do you think both have the same goal, @nikunjy

**Comment by nikunjy** [comment]
@LeoColman  No. That one is very specific to gradle. Here is a comment by @sksamuel  
https://github.com/kotest/kotest/issues/2416#issuecomment-1233571437

>I'm happy to expand the launcher but I'd rather do it better than the
gradle way (we just do that for compatibility).
>Perhaps accept a wildcard for --spec and --test or something ?


It remains to be seen if what I am doing here is the "better" way. But I am motivated to push this to completion. :) Happy to accomodate any comments and revisions.

**Comment by nikunjy** [comment]
@LeoColman one of my tests failed. I fixed it, it requires your approval to run these tests. Can you please retrigger them, sorry about that. Also I notice a lot of tests in this repo get skipped, is this expected ?

**Comment by jschneidereit** [comment]
I'm a little confused about what we're doing here - I'm pretty sure most of this functionality exists in these two files:

kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/spec/interceptor/SystemPropertySpecFilterInterceptor.kt

and

kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/status/SystemPropertyTestFilterEnabledExtension.kt

all you should have to do iirc is set the sysprop/envvar appropriately before starting your runner

**Comment by nikunjy** [comment]
@jschneidereit If this PR is merged you will be able to use the `kotest-framework-engine-launcer` with the following args 
```
"--reporter", "teamcity", 
"--package", "com.kotlin.test",
"--spec", "*",
"--testpath", "db*",
```
The launcher already supports `--testpath` argument. And this PR is just extending that `testpath` option to be more flexible. It is written with the same reasoning in mind as the original `--testpath` arg. 

Not only this but you would be able to use this to write a runner like this 
``` 
 ExecuteKt.execute(
  reporter, 
  "<package name>",
  "<spec wildcard>"
  "<test wildcard>"
)
```

There is some conversation in here https://github.com/kotest/kotest/issues/2416 as well

**Comment by nikunjy** [comment]
Changed the description. Thanks for asking @jschneidereit  I should added more description in the PR

**Comment by LeoColman** [comment]
I approved the run again.
>  No. That one is very specific to gradle

I stand corrected, I overlooked the "wildcard" keyword only :P

**Comment by nikunjy** [comment]
Looks like the tests have passed. What are the next steps @LeoColman

**Comment by jschneidereit** [comment]
What I was getting at is I think this is a duplicate implementation

**Comment by nikunjy** [comment]
@jschneidereit can you elaborate. I highlighted the reasons on why it is different. 
Furthermore the motivation came from conversation with the author @sksamuel 
He asked me to extend the launcher jar and in a different way than project does for gradle.

You can target test, subtest etc by providing the right parameters `--testpath foo * -- bar*` . Can you help me understand why you still think is duplicate.

**Comment by sksamuel** [comment]
I don't think the existing stuff supports the wildcard @jschneidereit ?

**Comment by jschneidereit** [comment]
I'm probably misunderstanding something here, but I recall adding test filtering by regex here:

https://github.com/kotest/kotest/pull/2547/files#diff-c8b53629477f4bcb438b5764976c6fc1d2999cf499bf53a7b062fdb986531de4R106

**Comment by nikunjy** [comment]
Would that approach apply if I am using the fat jar `kotest-framework-engine-launcer` binary (written [here](https://github.com/kotest/kotest/issues/2416)). If yes then I guess my question is why does the it take `--testpath` or `--spec` at all in that case.

**Comment by jschneidereit** [comment]
It might be that that fat jar author didn't know about the test/spec filter settings, or that I'm completely wrong 😂

**Comment by nikunjy** [comment]
I think this is reasonable extension to the far jar launcher unless you are considering removing it. I am looking forward to using the launcher binary. Can we please merge this since it builds on the existing functionality and I don't think it conflicts with gradle filter feature.

**Review by LeoColman — APPROVED** [review]

**Comment by sksamuel** [comment]
I think this feature is good because in the future the launcher will need to know tests before we get into the engine, as we build out our own gradle plugin. The existing filter code works, but only when the tests are known in advance (like in the current plugin).

**Comment by sksamuel** [comment]
We can merge this once green.

**Review by sksamuel — APPROVED** [review]

**Pull request merged**
